### PR TITLE
WIP: Initial ML-KEM support

### PR DESCRIPTION
**Pull request by mattcaswell**
This PR add initial support for ML-KEM.

It is based on the Kyber reference implementation from the "standard" branch here (which according the LICENSE file is available as "public domain" software):

https://github.com/pq-crystals/kyber

We will need to figure out whether we do about CLAs for this reference code. Most of the work in this PR is OpenSSL plumbing, so if we decide to incorporate a different implementation of ML-KEM instead much of the work here could be reused. Also the ref code has been incorporated "as is". It will need updating to OpenSSL coding standards.

This is a WIP. So far I have only integrated support for the "ref" code which is a pure C based implementation. There is also some code optimised for AVX2 available, which I have included in this PR but have not yet integrated - so it's not currently used.

The reference code can support ML-KEM-512, ML-KEM-768 and ML-KEM-1024 - but the way the code is structured you can only compile one of those at a time. So for this PR we only support ML-KEM-768. It will need some refactoring to support all 3 variants.

I have included a basic test, but more extensive testing will be required. There are some challenges to this due to the non-deterministic nature of the algorithm (although a deterministic variant is possible).

I have also not included any support for serialization/deserialization, or import/export from the provider.

So work left to do:
- Figure out what to do about CLAs
- Refactor the code to support ML-KEM-512, ML-KEM-768 and ML-KEM-1024 all at the same time
- Integrate the AVX2 optimised code
- Update the ref code to OpenSSL coding standards
- Extend the testing
- Add support for serialization/deserialization/import/export
- Refactor to use our own SHAKE implementation

With the above caveats and issues in mind, this does provide a somewhat useable ML-KEM implementation.

### CODE REVIEW

**Comment by mattcaswell** [comment]
Given the list of "TODOs" above it might be worth considering putting this on a feature branch.

**Comment by romen** [comment]
This is great work @mattcaswell !

To the list of TODOs I would probably add that we should probably also think about how to do the plumbing to support PQ/T Hybrid schemes (e.g., [X25519+ML-KEM768 or P256+ML-KEM768][ecdhe-mlkem]).

In EU, at least during the transition, hybrid solutions are a requirement.

[ecdhe-mlkem]: https://datatracker.ietf.org/doc/draft-kwiatkowski-tls-ecdhe-mlkem/

**Comment by romen** [comment]
Regarding CLA, https://github.com/pq-crystals/kyber/blob/main/AUTHORS list the authors.

Maybe @cryptojedi can help us collecting the necessary CLA forms?

**Comment by cryptojedi** [comment]
Nicola Tuveri ***@***.***> wrote:
> Regarding CLA, https://github.com/pq-crystals/kyber/blob/main/AUTHORS list the authors.
>
> Maybe @cryptojedi can help us collecting the necessary CLA forms?

Happy to help. What forms exactly do you need?

**Comment by paulidale** [comment]
The contributors lists is longer BTW: https://github.com/pq-crystals/kyber/graphs/contributors

File: crypto/mlkem/randombytes.c [hunk 1]
```diff
@@ -0,0 +1,80 @@
+#include <stddef.h>
```
Comment (paulidale) [inline crypto/mlkem/randombytes.c:1]: This entire file should be replaced by calls to `RAND_bytes` or `RAND_priv_bytes`

File: providers/defltprov.c [hunk 2]
```diff
@@ -477,6 +477,7 @@ static const OSSL_ALGORITHM deflt_asym_kem[] = {
 # endif
     { PROV_NAMES_EC, "provider=default", ossl_ec_asym_kem_functions },
 #endif
+    { PROV_NAMES_MLKEM, "provider=default", ossl_mlkem_asym_key_functions },
```
Comment (paulidale) [inline providers/defltprov.c:480]: Similar should be added to the FIPS provider.

File: crypto/mlkem/avx2/keccak4x/KeccakP-brg_endian.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
+/*
```
Comment (paulidale) [inline crypto/mlkem/avx2/keccak4x/KeccakP-brg_endian.h:1]: Do we really want a new Keccak implementation included?

File: crypto/mlkem/avx2/fips202.c [hunk 4]
```diff
@@ -0,0 +1 @@
+../ref/fips202.c
```
Comment (paulidale) [inline crypto/mlkem/avx2/fips202.c:1]: This file should be deduped and the work done in `build.info` I suspect.

**Review by paulidale — COMMENTED** [review]
Just highlighting a few items that ought to go onto the to do list.

File: providers/defltprov.c (hunk 2 above)
Comment (slontis) [reply providers/defltprov.c:480]: I would leave this out initially, as it also requires self tests before adding

**Comment by mattcaswell** [comment]
> Happy to help. What forms exactly do you need?

@cryptojedi - thanks for the offer of help. What we need is CLAs in place for anyone that has made any commits to the "standard" branch of the reference implementation.

The CLA forms are here:

https://openssl-library.org/policies/cla/index.html

Once completed they need to be sent to [cla@openssl.org](mailto:cla@openssl.org) (please reference that they are in relation to this PR). If the commits were made on behalf of an employer then both a CCLA and ICLA are required. Otherwise just an ICLA is enough. If sending ICLAs only you will be asked to confirm that the commits were made individually and not on behalf of an employer.

Anything you can do to help us collect the necessary paperwork would be greatly appreciated.

**Comment by mattcaswell** [comment]
> To the list of TODOs I would probably add that we should probably also think about how to do the plumbing to support PQ/T Hybrid schemes (e.g., [X25519+ML-KEM768 or P256+ML-KEM768](https://datatracker.ietf.org/doc/draft-kwiatkowski-tls-ecdhe-mlkem/)).

Yes - agreed. I didn't want to look too far ahead at this stage. The initial problem is getting an ML-KEM implementation at all!!

**Comment by mattcaswell** [comment]
This PR no longer seems relevant. See the feature/ml-kem branch for the current state of ML-KEM development.

**Pull request closed without merging**